File: Solution/Python/565. Array Nesting.py
class Solution:
    def arrayNesting(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        seen, res = [0] * len(nums), 0
        for i in nums:
            count, j = 0, i
            while not seen[j]:
                seen[j], count, j = 1, count + 1, nums[j]
            res = max(res, count)
        return res


s = Solution()
# res = s.arrayNesting([5, 4, 0, 3, 1, 6, 2])
res = s.arrayNesting([1, 0, 5, 4, 0, 3, 1, 6, 2])
print(res)

File: Solution/Javascript/565. Array Nesting.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var arrayNesting = function (nums) {
    let res = 0, set = new Set();
    for (let i = 0; i < nums.length; i++) {
        let curIndex = i, count = 0;
        while (!set.has(curIndex)) {
            set.add(curIndex);
            curIndex = nums[curIndex];
            count++;
        }
        res = Math.max(res, count);
    }
    return res;
};

let arr = [5, 4, 0, 3, 1, 6, 2];
console.log(arrayNesting(arr));